master-firmware/src/strategy/color_sequence_server.c

```diff
@@ -0,0 +1,67 @@
+#include <ch.h>
+#include <hal.h>
+
+#include <error/error.h>
+
+#include "main.h"
+#include "priorities.h"
+
+#include "robot_helpers/eurobot2018.h"
+#include "color_sequence_server.h"
+
+#define COLOR_SEQUENCE_SERVER_STACKSIZE 512
+
+// Bluetooth UART
+#define BLUETOOTH_UART_BAUDRATE 19200
+static const SerialConfig bt_uart_config = {
+    .speed = BLUETOOTH_UART_BAUDRATE,
+    .cr1 = 0,
+    .cr2 = USART_CR2_STOP1_BITS | USART_CR2_LINEN,
+    .cr3 = 0
+};
+
+static THD_FUNCTION(color_sequence_server_thd, arg) {
+    (void)arg;
+    chRegSetThreadName(__FUNCTION__);
+
+    uint8_t buffer[16];
+    sdStart(&SD2, &bt_uart_config);
+
+    static enum cube_color colors[5] = {
+        CUBE_UNKNOWN, CUBE_UNKNOWN, CUBE_UNKNOWN, CUBE_UNKNOWN, CUBE_UNKNOWN
+    };
+    static messagebus_topic_t colors_topic;
+    static MUTEX_DECL(colors_topic_lock);
+    static CONDVAR_DECL(colors_topic_condvar);
+    messagebus_topic_init(&colors_topic, &colors_topic_lock, &colors_topic_condvar, &colors[0], sizeof(colors));
+    messagebus_advertise_topic(&bus, &colors_topic, "/colors");
+
+    NOTICE("Color sequence server up, listening over BT");
+
+    while (true) {
+        size_t len = sdAsynchronousRead(&SD2, buffer, sizeof(buffer));
+
+        if (len < 3) {
```

Comment:
What do you think of my suggestion (above?)

Reply:
:scream: how does it even work

Reply:
Thats what code reviews are for 👼

Reply:
I think you're filling the buffer when you read `len >= 3` characters, which is then processed in following loop when `len == 0`.

Reply:
@antoinealb I suggest to discard the data if the line has not the correct length.

Reply:
... because the bluetooth modules sends something like `"Connect to: 0012-6F-22FEDD\n"` upon connection.

Reply:
But we only consider valid characters (Y,B,K,G,O) in `color_sequence_server`, otherwise the color is set to unknown, and the strategy would consider that nothing was received from the vision beacon. So we should be safe without length check.

Reply:
I think the best would be to change the server to send something like `@YBK\n` (start and end marker) so you can sync on the @ and then read 4 and check that `buffer[3] == '\n'`.

Of course better would be to run TCP/IP over serial link and stuff but hey Aint nobody got time for this.

Reply:
also, the color sequence server only starts after we select the color side on the bot, so the BT module should have booted by then.

Reply:
I'm currently adding the sync character. The checksum is left for the reader as an exercice :trollface:

Reply:
> also, the color sequence server only starts after we select the color side on the bot, so the BT module should have booted by then.

Totally the right way to build reliable software 😉